modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java

```diff
@@ -296,6 +312,115 @@ public static CompletableFuture<?>[] triggerAllTablePartitionsRebalance(
+
+            futures[partId] = tableAssignmentsFut.thenCompose(tableAssignments ->
+                    tableAssignments.isEmpty() ? nullCompletedFuture() : manualPartitionUpdate(
+                            replicaGrpId,
+                            dataNodes,
+                            aliveNodesConsistentIds,
+                            zoneDescriptor.replicas(),
+                            revision,
+                            metaStorageManager,
+                            tableAssignments.get(replicaGrpId.partitionId()).nodes()
+                    )).thenAccept(res -> {
+                        LOG.info("Partition {} returned {} status on reset attempt", replicaGrpId, UpdateStatus.valueOf(res));
+                    }
+            );
+        }
+
+        return futures;
+    }
+
+    private static CompletableFuture<Integer> manualPartitionUpdate(
+            TablePartitionId partId,
+            Collection<String> dataNodes,
+            Set<String> aliveNodesConsistentIds,
+            int replicas,
+            long revision,
+            MetaStorageManager metaStorageMgr,
+            Set<Assignment> currentAssignments
+    ) {
+        Set<Assignment> calcAssignments = AffinityUtils.calculateAssignmentForPartition(dataNodes, partId.partitionId(), replicas);
+
+        // TODO https://issues.apache.org/jira/browse/IGNITE-21303
+        //  This is a naive approach that doesn't exclude nodes in error state, if they exist.
+        Set<Assignment> partAssignments = new HashSet<>();
+        for (Assignment currentAssignment : currentAssignments) {
+            if (aliveNodesConsistentIds.contains(currentAssignment.consistentId())) {
+                partAssignments.add(currentAssignment);
+            }
+        }
+
+        if (partAssignments.size() >= (replicas / 2 + 1)) {
+            return CompletableFuture.completedFuture(ASSIGNMENT_NOT_UPDATED.ordinal());
+        }
+
+        for (Assignment calcAssignment : calcAssignments) {
+            if (partAssignments.size() == replicas) {
+                break;
+            }
+
+            partAssignments.add(calcAssignment);
+        }
+
+        if (currentAssignments.equals(calcAssignments)) {
+            return CompletableFuture.completedFuture(ASSIGNMENT_NOT_UPDATED.ordinal());
+        }
+
+        byte[] partAssignmentsBytes = Assignments.forced(partAssignments).toBytes();
+        byte[] revisionBytes = ByteUtils.longToBytes(revision);
+
+        ByteArray partChangeTriggerKey = pendingChangeTriggerKey(partId);
+        ByteArray partAssignmentsPendingKey = pendingPartAssignmentsKey(partId);
+
+        Iif iif = iif(
+                notExists(partChangeTriggerKey).or(value(partChangeTriggerKey).lt(revisionBytes)),
+                iif(
+                        value(partAssignmentsPendingKey).ne(partAssignmentsBytes),
+                        ops(
+                                put(partAssignmentsPendingKey, partAssignmentsBytes),
```

Comment:
I wonder if we need to erase planned key when we manually have changed pending?

Reply:
Yes we do